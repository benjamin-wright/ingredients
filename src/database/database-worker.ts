import initSqlJs from '@jlongster/sql.js'
import { SQLiteFS } from 'absurd-sql'
import IndexedDBBackend from 'absurd-sql/dist/indexeddb-backend'
import absurdSqlWasmUrl from '@jlongster/sql.js/dist/sql-wasm.wasm?url'
import { type ExecRequest } from './request-types'

async function run() {
  const SQL = await initSqlJs({ locateFile: () => absurdSqlWasmUrl })
  const sqlFS = new SQLiteFS(SQL.FS, new IndexedDBBackend())
  SQL.register_for_idb(sqlFS)

  SQL.FS.mkdir('/sql')
  SQL.FS.mount(sqlFS, {}, '/sql')

  const path = '/sql/db.sqlite'
  if (typeof SharedArrayBuffer === 'undefined') {
    const stream = SQL.FS.open(path, 'a+')
    await stream.node.contents.readIfFallback()
    SQL.FS.close(stream)
  }

  const db = new SQL.Database(path, { filename: true })
  // You might want to try `PRAGMA page_size=8192;` too!
  db.exec(`
    PRAGMA journal_mode=MEMORY;
    PRAGMA page_size=8192;
  `)

  // listen for messages
  onmessage = async (event: MessageEvent<any>) => {
    switch (event.data.type) {
      case 'exec':
        onExec(db, event.data)
        break
      default:
        console.error('Unknown message type', event.data)
    }
  }
}

function execResult(id: number, data: any) {
  postMessage({
    type: 'exec',
    requestId: id,
    result: data,
  })
}

function execError(id: number, error: string) {
  postMessage({
    type: 'exec',
    requestId: id,
    error,
  })
}

function onExec(db: any, request: ExecRequest) {
  try {
    db.exec('BEGIN TRANSACTION');
    const rows = db.exec(
      request.sql,
      request.bind
    );
    db.exec('COMMIT');
    execResult(request.requestId, rows);
  } catch (err: any) {
    execError(request.requestId, err.message);
  }
}

run()
