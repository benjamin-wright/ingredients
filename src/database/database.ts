import { initBackend } from 'absurd-sql/dist/indexeddb-main-thread';
import hash from '@/utils/hash';
import DBWorker from './database-worker.ts?worker';
import { type ExecResponse, type ExecRequest } from './request-types';

import migration00 from './migrations/migration-00.sql?raw';
import migration01 from './migrations/migration-01.sql?raw';
import migration02 from './migrations/migration-02.sql?raw';
import migration03 from './migrations/migration-03.sql?raw';

type Request<T> = {
  resolve: (value: T[]) => void;
  reject: (reason?: any) => void;
  reader?: (row: any[]) => T;
}

class Database {
  private queryCounter = 0;
  private worker: Worker;
  private initPromise;
  private initResolver: () => void;
  private requests: Record<number, Request<any>> = {};

  constructor() {
    this.initResolver = () => {};
    this.initPromise = new Promise<void>((resolve) => {
      this.initResolver = resolve;
    });

    this.worker = new DBWorker();
    this.worker.onmessage = this.receiveMessage.bind(this);
    this.worker.onerror = this.receieveError.bind(this);

    initBackend(this.worker);
  }

  async init(): Promise<void> {
    console.log('waiting for init...');
    await this.initPromise;
    console.log('ready');
  }

  private receiveMessage(event: MessageEvent<any>) {
    switch (event.data.type) {
      case 'exec': {
        const response = event.data as ExecResponse;
        const request = this.requests[response.requestId];

        if (response.error) {
          request.reject(response.error);
          return;
        }

        if (!request) {
          console.error('Unknown request id', response.requestId);
          return;
        }

        if (response.result) {
          const output: any[][] = [];
          if (request.reader) {
            for (const result of response.result) {
              for (const row of result.values) {
                output.push(request.reader(row));
              }
            }
          } else {
            for (const result of response.result) {
              output.push(...result.values);
            }
          }
          request.resolve(output);
        } else {
          request.resolve([]);
        }

        break;
      }
      case '__absurd:spawn-idb-worker': {
        this.initResolver();
        break;
      }
      default: {
        console.error('Unknown message type', event.data);
      }
    }
  }

  private receieveError(event: ErrorEvent) {
    console.error(event);
  }

  async reset(): Promise<void> {
    this.worker.postMessage({ type: 'close' });
    const request = window.indexedDB.deleteDatabase('db.sqlite');
    return new Promise((resolve, reject) => {
      request.onsuccess = () => resolve();
      request.onerror = () => reject(new Error('Error deleting database'));
      request.onblocked = () => reject(new Error('Blocked deleting database'));
    });
  }

  async migrate(migrations: string[]): Promise<void> {
    console.info('Initializing migrations...');
    await this.initMigrations();

    for (const [n, sql] of migrations.entries()) {
      if (!await this.hasMigration(n, sql)) {
        console.info(`Running migration ${n}...`);
        await this.runMigration(n, sql);
      } else {
        console.info(`Skipping migration ${n}...`);
      }
    }

    console.info('Done migrating.');
  }

  exec<T>(sql: string, bind?: any[], reader?: (row: any[]) => T): Promise<T[]> {
    const id = this.queryCounter++;
    return new Promise<T[]>((resolve, reject) => {
      this.requests[id] = { resolve, reject, reader };
      const request: ExecRequest = { type: 'exec', requestId: id, sql, bind };
      this.worker.postMessage(request);
    });
  }

  private async initMigrations(): Promise<void> {
    await this.exec('CREATE TABLE IF NOT EXISTS migrations (id INTEGER PRIMARY KEY, hash INTEGER NOT NULL);');
  }

  private async hasMigration(n: number, sql: string): Promise<boolean> {
    const rows = await this.exec(
      'SELECT id, hash FROM migrations WHERE id == ?',
      [n],
      (row) => ({ id: row[0] as number, hash: row[1] as number }),
    );

    if (rows.length == 0) {
      return false;
    }

    const row = rows[0];

    if (row.hash != hash(sql)) {
      throw new Error(
        `Migration ${n} has already been run, but the hash does not match.`
      );
    }

    return true;
  }

  private async runMigration(n: number, sql: string): Promise<void> {
    await this.exec(sql);
    await this.exec(
      'INSERT INTO migrations (id, hash) VALUES (?, ?)',
      [n, hash(sql)],
    );
  }
}

const database = (async () => {
  const db = new Database();
  await db.init();
  return db;
})();

const migrated = database.then(async (db) => {
  await db.migrate([migration00, migration01, migration02, migration03]);
  return db;
});

export async function query<T>(sql: string, bind?: any[], reader?: (row: any[]) => T): Promise<T[]> {
  return migrated.then(db => db.exec(sql, bind, reader));
}

export async function reset(): Promise<void> {
  return database.then(db => db.reset());
}
