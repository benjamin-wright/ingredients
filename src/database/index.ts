export * from './database';
// import Database from './database';

// import migration0 from './migration-0.sql?raw';

// export async function GetDB(): Promise<Database> {
//     const sqlite = await Database.getSqliteDB();
//     const db = new Database(sqlite);

//     db.init([migration0]);

//     return db;
// } 