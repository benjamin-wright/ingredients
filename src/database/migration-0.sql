CREATE TABLE "categories" {
    "id" INTEGER PRIMARY KEY AUTOINCREMENT,
    "name" TEXT NOT NULL,
    "created_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updated_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
};

CREATE TABLE "quantity_units" {
    "id" INTEGER PRIMARY KEY AUTOINCREMENT,
    "name" TEXT NOT NULL,
    "singular" TEXT NOT NULL,
    "plural" TEXT NOT NULL,
    "is_unit" BOOLEAN NOT NULL DEFAULT 0,
    "created_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updated_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
};

CREATE TABLE "ingredients" {
    "id" INTEGER PRIMARY KEY AUTOINCREMENT,
    "name" TEXT NOT NULL,
    "category_id" INTEGER NOT NULL,
    "created_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updated_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    FOREIGN KEY ("category_id") REFERENCES "categories" ("id") ON DELETE CASCADE
};

CREATE TABLE "recipies" {
    "id" INTEGER PRIMARY KEY AUTOINCREMENT,
    "name" TEXT NOT NULL,
    "description" TEXT NOT NULL,
    "created_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updated_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
};

CREATE TABLE "recipie_ingredients" {
    "id" INTEGER PRIMARY KEY AUTOINCREMENT,
    "recipie_id" INTEGER NOT NULL,
    "ingredient_id" INTEGER NOT NULL,
    "quantity" REAL NOT NULL,
    "created_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updated_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    FOREIGN KEY ("recipie_id") REFERENCES "recipies" ("id") ON DELETE CASCADE,
    FOREIGN KEY ("ingredient_id") REFERENCES "ingredients" ("id") ON DELETE CASCADE
};

CREATE TABLE "recipie_steps" {
    "id" INTEGER PRIMARY KEY AUTOINCREMENT,
    "recipie_id" INTEGER NOT NULL,
    "step" INTEGER NOT NULL,
    "description" TEXT NOT NULL,
    "created_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updated_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    FOREIGN KEY ("recipie_id") REFERENCES "recipies" ("id") ON DELETE CASCADE
};

CREATE TABLE "dinner_plans" {
    "id" INTEGER PRIMARY KEY AUTOINCREMENT,
    "day" INTEGER,
    "recipie_id" INTEGER NOT NULL,
    "created_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updated_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    FOREIGN KEY ("recipie_id") REFERENCES "recipies" ("id") ON DELETE CASCADE
};

CREATE TABLE "non_dinner_plans" {
    "id" INTEGER PRIMARY KEY AUTOINCREMENT,
    "people" INTEGER,
    "portions" INTEGER,
    "recipie_id" INTEGER NOT NULL,
    "description" TEXT NOT NULL,
    "created_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updated_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    FOREIGN KEY ("recipie_id") REFERENCES "recipies" ("id") ON DELETE CASCADE
};

CREATE TABLE "shopping_list_items" {
    "id" INTEGER PRIMARY KEY AUTOINCREMENT,
    "ingredient_id" INTEGER NOT NULL,
    "quantity" REAL NOT NULL,
    "quantity_unit_id" INTEGER NOT NULL,
    "got" BOOLEAN NOT NULL DEFAULT 0,
    "created_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updated_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    FOREIGN KEY ("ingredient_id") REFERENCES "ingredients" ("id") ON DELETE CASCADE,
    FOREIGN KEY ("quantity_unit_id") REFERENCES "quantity_units" ("id") ON DELETE CASCADE
};

CREATE TABLE "shopping_list_custom_items" {
    "id" INTEGER PRIMARY KEY AUTOINCREMENT,
    "name" TEXT NOT NULL,
    "quantity" REAL NOT NULL,
    "quantity_unit_id" INTEGER NOT NULL,
    "got" BOOLEAN NOT NULL DEFAULT 0,
    "created_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updated_at" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    FOREIGN KEY ("quantity_unit_id") REFERENCES "quantity_units" ("id") ON DELETE CASCADE
};
