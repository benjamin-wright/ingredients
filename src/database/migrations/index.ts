import migration00 from './migration-00.sql?raw';
import migration01 from './migration-01.sql?raw';
import migration02 from './migration-02.sql?raw';
import migration03 from './migration-03.sql?raw';

export default [
    migration00,
    migration01,
    migration02,
    migration03,
]