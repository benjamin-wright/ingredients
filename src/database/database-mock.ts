import BetterSqlite3 from 'better-sqlite3';
import migrations from './migrations';

export class MockDatabase {
  private db: BetterSqlite3.Database;
  
  constructor() {
    this.db = new BetterSqlite3(':memory:');
    this.db.exec(`
        PRAGMA journal_mode=MEMORY;
        PRAGMA page_size=8192;
        PRAGMA foreign_keys=ON;
    `);

    migrations.forEach(m => this.db.exec(m));
  }

  async query<T>(sql: string, bind?: any[], reader?: (row: any[]) => T): Promise<T[]> {
    let statement = this.db.prepare<any[]>(sql)
    
    if (bind) {
      statement = statement.bind(bind);
    }

    if (reader) {
      return statement.raw(true).all().map<any>((value: unknown) => reader(value as any[]));
    }

    statement.run();
    return [];
  }
}