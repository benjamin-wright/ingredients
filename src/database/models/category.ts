import { query } from "@/database/database";

export type Category = {
    id: number;
    position: number;
    name: string;
}

export async function getCategories(): Promise<Category[]> {
    const categories = await query(
        /*sql*/`SELECT id, position, name FROM categories ORDER BY position ASC`,
        [], (values) => {
            return {
                id: values[0] as number,
                position: values[1] as number,
                name: values[2] as string,
            };
        }
    );

    return categories;
}

export async function deleteCategory(id: number): Promise<void> {
    await query(
        /*sql*/`DELETE FROM categories WHERE id = ?`,
        [id]
    );
}

export async function addCategory(name: string): Promise<number> {
    let result = await query(
        /*sql*/`SELECT MAX(position) FROM categories`,
        [],
        (values) => values[0] as number
    );
    const position = result[0] + 1 ?? 1;

    result = await query(
        /*sql*/`INSERT INTO categories (name, position) VALUES (?, ?) RETURNING id`,
        [name, position],
        (values) => values[0] as number
    );

    return result[0];
}

export async function updateCategory(id: number, name: string): Promise<void> {
    await query(
        /*sql*/`UPDATE categories SET name = ? WHERE id = ?`,
        [name, id]
    );
}

export async function swapCategories(a: Category, b: Category): Promise<void> {
    await query(
        /*sql*/`UPDATE categories SET position = ? WHERE id = ?`,
        [b.position, a.id]
    );
    await query(
        /*sql*/`UPDATE categories SET position = ? WHERE id = ?`,
        [a.position, b.id]
    );
}