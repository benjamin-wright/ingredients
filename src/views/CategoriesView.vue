<script setup lang="ts">
  import { onMounted, ref } from "vue";
  import { useRouter } from 'vue-router';
  import ObjectList from "../components/ObjectList.vue";
  import NewThing from "@/components/NewThing.vue";
  import { getCategories, deleteCategory, swapCategories, type Category } from "@/database/models/category";
  import { useNewCategoryStore } from "@/stores/new-category";

  const router = useRouter();
  const store = useNewCategoryStore();

  const loading = ref(true);
  const categories = ref([] as Category[]);

  onMounted(async () => {
    categories.value = await getCategories();
    loading.value = false;
  });

  async function remove(category: Category) {
    await deleteCategory(category.id);
    categories.value.splice(categories.value.indexOf(category), 1);
  }

  function edit(category: Category) {
    store.select(category);
    router.push("/categories/new");
  }

  async function swap(category1: Category, category2: Category) {
    await swapCategories(category1, category2);
    categories.value = await getCategories();
  }
</script>

<template>
  <main>
    <template v-if="loading">Loading...</template>
    <template v-else>
      <h1>Categories</h1>
      <ObjectList
        :data="categories"
        @delete="remove"
        @edit="edit"
        @swap="swap"
        reorder
      >
        <template #content="content">
          <h2>{{ content.obj.name }}</h2>
        </template>
      </ObjectList>
      <NewThing to="/categories/new" />
    </template>
  </main>
</template>

<style scoped>
  h2 {
    overflow: hidden;
    text-overflow: ellipsis;
  }
</style>