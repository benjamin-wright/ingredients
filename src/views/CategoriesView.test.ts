import { vi, describe, test, beforeEach, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import CategoriesView from './CategoriesView.vue';

import { MockDatabase } from '@/database/database-mock';
import { addCategory } from '@/database/models/category';
let mockDatabase: MockDatabase;

vi.mock('@/database/database', async () => {
    return {
        query: async (sql: string, bind?: any[], reader?: (row: any[]) => any) => {
            return mockDatabase.query(sql, bind, reader);
        }
    }
});

vi.mock('vue-router', () => ({
    useRouter: () => ({
        push: async () => {},
    }),
}));

describe('CategoriesView', () => {
    beforeEach(async () => {
        mockDatabase = new MockDatabase();
    });

    test('should render', async () => {
        await addCategory('Test');

        const wrapper = mount(CategoriesView, {
            global: {
                stubs: ['FontAwesomeIcon'],
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });
});