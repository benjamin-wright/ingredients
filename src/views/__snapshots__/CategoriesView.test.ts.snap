// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CategoriesView > should render 1`] = `"<main data-v-6a11679b=\\"\\">Loading...</main>"`;
